cli/command.go

```diff
@@ -0,0 +1,282 @@
+	SetStdin(r io.Reader)
+
+	// Pipe creates new unqiue stdin, stdout, and stderr buffers, sets them on the
+	// command, and returns them. This is most useful for testing where callers
+	// want to simulate inputs or assert certain command outputs.
+	Pipe() (stdin, stdout, stderr *bytes.Buffer)
+}
+
+// CommandFactory returns a new instance of a command. This returns a function
+// instead of allocations because we want the CLI to load as fast as possible,
+// so we lazy load as much as possible.
+type CommandFactory func() Command
+
+// Ensure [RootCommand] implements [Command].
+var _ Command = (*RootCommand)(nil)
+
+// RootCommand represents a command root for a parent or collection of
+// subcommands.
+type RootCommand struct {
+	BaseCommand
+
+	// Name is the name of the command or subcommand. For top-level commands, this
+	// should be the binary name. For subcommands, this should be the name of the
+	// subcommand.
+	Name string
+
+	// Description is the human-friendly description of the command.
+	Description string
+
+	// Hide marks the entire subcommand as hidden. It will not be shown in help
+	// output.
+	Hide bool
+
+	// Version defines the version information for the command. This can be
+	// omitted for subcommands as it will be inherited from the parent.
+	Version string
+
+	// Commands is the list of sub commands.
+	Commands map[string]CommandFactory
+}
+
+// Desc is the root command description. It is used to satisfy the [Command]
+// interface.
+func (r *RootCommand) Desc() string {
+	return r.Description
+}
+
+// Hidden determines whether the command group is hidden. It is used to satisfy
+// the [Command] interface.
+func (r *RootCommand) Hidden() bool {
+	return r.Hide
+}
+
+// Help compiles structured help information. It is used to satisfy the
+// [Command] interface.
+func (r *RootCommand) Help() string {
+	var b strings.Builder
+
+	longest := 0
+	names := make([]string, 0, len(r.Commands))
+	for name := range r.Commands {
+		names = append(names, name)
+		if l := len(name); l > longest {
+			longest = l
+		}
+	}
+	sort.Strings(names)
+
+	fmt.Fprintf(&b, "Usage: %s COMMAND\n\n", r.Name)
+	for _, name := range names {
+		cmd := r.Commands[name]()
+		if cmd == nil {
+			continue
+		}
+
+		if !cmd.Hidden() {
+			fmt.Fprintf(&b, "  %-*s%s\n", longest+4, name, cmd.Desc())
+		}
+	}
+
+	return strings.TrimRight(b.String(), "\n")
+}
+
+// Run executes the command and prints help output or delegates to a subcommand.
+func (r *RootCommand) Run(ctx context.Context, args []string) error {
+	name, args := extractCommandAndArgs(args)
+
+	// Short-circuit top-level help.
+	if name == "" || name == "-h" || name == "-help" || name == "--help" {
+		fmt.Fprintln(r.Stderr(), r.Help())
+		return nil
+	}
+
+	// Short-circuit version.
+	if name == "-v" || name == "-version" || name == "--version" {
+		fmt.Fprintln(r.Stderr(), r.Version)
+		return nil
+	}
+
+	cmd, ok := r.Commands[name]
+	if !ok {
+		return fmt.Errorf("unknown command %q", name)
```

Comment:
nit: on unknown command mention the user can do -help to get a list of valid commands